[Pull Request] Add ROS2 Visualization for KISS-Matcher

kimdaebeom opened the pull request:
## Summary
This PR adds a **ROS2-based point cloud visualization** for KISS-Matcher registration results. The visualization animates the transformation of a source point cloud to match a target point cloud in RViz. `README.md` file is also updated for convenience.

LimHyungTae commented on ros/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+find_package(pcl_conversions REQUIRED)
+find_package(PCL REQUIRED)
+find_package(Eigen3 REQUIRED)
+find_package(teaserpp REQUIRED)
```
Do we need teaserpp library for this visualization?

LimHyungTae commented on ros/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+find_package(robin REQUIRED)
+find_package(kiss_matcher REQUIRED)
+
+add_executable(pointcloud_visualizer src/pointcloud_visualizer.cpp)
```
`reg_visualizer` would be a more clear name.

LimHyungTae commented on ros/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+  pcl_conversions
+  PCL
+  Eigen3
+  teaserpp
```
It would be nicer if you remove the dependency of `teaserpp`

LimHyungTae commented on ros/CMakeLists.txt:
```diff
@@ -0,0 +1,46 @@
+)
+target_link_libraries(pointcloud_visualizer
+  Eigen3::Eigen
+  teaserpp::teaser_registration
```
It would be nicer if you remove the dependency of `teaserpp`

LimHyungTae commented on ros/README.md:
```diff
@@ -0,0 +1,116 @@
+This repository provides a ROS2-based visualization tool for KISS-Matcher registration results.
+It allows users to animate the transformation of a source point cloud to match a target point cloud.
+
+## :package: Prerequisites
```
Since ROS2 automatically installs `pcl`, we don't need this Prerequisites part.

LimHyungTae commented on ros/README.md:
````diff
@@ -0,0 +1,116 @@
+
+```bash
+cd ${ROS2_WORKSPACE}
+colcon build --packages-select pointcloud_visualizer
````
Again, could you change the name to `reg_visualizer`, `matching_visualizer`, or `registration_visualizer`?

LimHyungTae replied:
Because it is not visualizing the point cloud itself but the registration process.

LimHyungTae commented on ros/README.md:
```diff
@@ -0,0 +1,116 @@
+
+## 🛠 Configuration
+
+You can customize the visualization parameters in **`config/params.yaml`** before launching the node.
```
Could you add the mention that one can download data using [download_datasets.py](https://github.com/kimdaebeom/KISS-Matcher/blob/main/python/utils/download_datasets.py) easily?

LimHyungTae commented on ros/README.md:
```diff
@@ -0,0 +1,116 @@
+    BASE_DIR: "src/KISS-Matcher/cpp/examples/build/data"
+
+    # Specify the source and target PCD files
+    SRC_PCD_DIR: "Vel64/kitti_000540.pcd"
```
In my opinion, 'DIR' usually indicates a folder directory. So, could you change the name to `src_pcd_path` or `src_pcd_file` for clarity?

LimHyungTae replied:
+

LimHyungTae commented on ros/README.md:
```diff
@@ -0,0 +1,116 @@
+
+    # Specify the source and target PCD files
+    SRC_PCD_DIR: "Vel64/kitti_000540.pcd"
+    TGT_PCD_DIR: "Vel64/kitti_001319.pcd"
```
In my opinion, 'DIR' usually indicates a folder directory. So, could you change the name to `tgt_pcd_path` or `tgt_pcd_file` for clarity?

LimHyungTae commented on ros/README.md:
```diff
@@ -0,0 +1,116 @@
+
+| Parameter      | Description |
+|---------------|-------------|
+| `SRC_PCD_DIR`  | Path to the source PCD file |
+| `TGT_PCD_DIR`  | Path to the target PCD file |
```
please fix these lines as well!

LimHyungTae commented on ros/config/params.yaml:
```diff
@@ -0,0 +1,22 @@
+pointcloud_visualizer:
+  ros__parameters:
+    base_dir: "src/KISS-Matcher/cpp/examples/build/data/"
+    # src_pcd: "HeLiPR-KAIST05/Aeva.pcd"
+    # tgt_pcd: "HeLiPR-KAIST05/Ouster.pcd"
+    # src_pcd: "KITTI00-to-07/kitti00.pcd"
+    # tgt_pcd: "KITTI00-to-07/kitti07.pcd"
+    # src_pcd: "KITTI00-to-KITTI360/kitti00.pcd"
+    # tgt_pcd: "KITTI00-to-KITTI360/kitti360_09.pcd"
+    # src_pcd: "VBR-Collosseo/test0.pcd"
+    # tgt_pcd: "VBR-Collosseo/train0.pcd"
+    src_pcd: "Vel16/src.pcd"
+    tgt_pcd: "Vel16/tgt.pcd"
+    # src_pcd: "Vel64/kitti_000540.pcd"
+    # tgt_pcd: "Vel64/kitti_001319.pcd"
+    # src_pcd: "VBR-Collosseo/test0.pcd"
+    # tgt_pcd: "VBR-Collosseo/train0.pcd"
```
This is great! At the top of this code block, could you add a comment line saying, "NOTE(dkim): Thes are provided examples of PCD pairs"?

LimHyungTae commented on ros/src/pointcloud_visualizer.cpp:
```diff
@@ -0,0 +1,188 @@
+    pcl::transformPointCloud(*source_cloud_, *source_cloud_, scale_transform);
+    pcl::transformPointCloud(*target_cloud_, *target_cloud_, scale_transform);
+
+    downsampleCloud(source_cloud_);
+    downsampleCloud(target_cloud_);
```
Now KISS-Matcher has downsampling itself! We don't need downsampling in advance. Please see here: 

```
  auto processInput = [&](const std::vector<Eigen::Vector3f> &input_cloud) {
    if (config_.use_voxel_sampling_) {
      return VoxelgridSampling(input_cloud, config_.voxel_size_);
    }
    return input_cloud;
  };
```

https://github.com/MIT-SPARK/KISS-Matcher/blob/f815c6fc20925def133b4c222cff3445d8900b8b/cpp/kiss_matcher/core/kiss_matcher/KISSMatcher.cpp#L48

LimHyungTae commented on ros/src/pointcloud_visualizer.cpp:
```diff
@@ -0,0 +1,188 @@
+
+    Eigen::Matrix4f accumulated_transform = Eigen::Matrix4f::Identity();
+
+    for (double step = 0; step <= 2 * num_steps_; ++step) {
+      pcl::PointCloud<pcl::PointXYZ>::Ptr animated_cloud =
+          std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
+      Eigen::Matrix4f frame_transform = Eigen::Matrix4f::Identity();
+
+      if (step <= num_steps_) {
```
This part can be more clear. For clarity, could you please change `2 * num_steps_` to `num_total_steps_` and `num_steps` below to `num_total_steps_ / 2`?

LimHyungTae commented on ros/src/pointcloud_visualizer.cpp:
```diff
@@ -0,0 +1,188 @@
+  }
+
+ private:
+  std::string base_path_, source_pcd_path_, target_pcd_path_;
+  double voxel_size_, num_steps_, animation_fps_, scale_factor_;
+  Eigen::Matrix4f estimated_transform_ = Eigen::Matrix4f::Identity();
```
Could you place each variable in each line?

LimHyungTae commented on ros/src/pointcloud_visualizer.cpp:
```diff
@@ -0,0 +1,188 @@
+  }
+
+ private:
+  std::string base_path_, source_pcd_path_, target_pcd_path_;
+  double voxel_size_, num_steps_, animation_fps_, scale_factor_;
+  Eigen::Matrix4f estimated_transform_ = Eigen::Matrix4f::Identity();
+
+  pcl::PointCloud<pcl::PointXYZ>::Ptr source_cloud_ =
+      std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
+  pcl::PointCloud<pcl::PointXYZ>::Ptr target_cloud_ =
+      std::make_shared<pcl::PointCloud<pcl::PointXYZ>>();
+
+  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr source_publisher_, target_publisher_;
+
+  void declareAndGetParameters() {
+    declare_parameter("base_dir", "");
+    declare_parameter("src_pcd", "");
+    declare_parameter("tgt_pcd", "");
+    declare_parameter("resolution", 0.1);
+    declare_parameter("moving_rate", 10.0);
+    declare_parameter("frame_rate", 10.0);
+    declare_parameter("scale", 1.0);
+
+    get_parameter("base_dir", base_path_);
+    get_parameter("src_pcd", source_pcd_path_);
+    get_parameter("tgt_pcd", target_pcd_path_);
+    get_parameter("resolution", voxel_size_);
+    get_parameter("moving_rate", num_steps_);
+    get_parameter("frame_rate", animation_fps_);
+    get_parameter("scale", scale_factor_);
```
In ROS2, is it not allowed for variables and parameters to have the same name? If it is possible, it would be better to name the variable the same as the key in the input YAML for clarity.

LimHyungTae reviewed (commented):
Thanks @kimdaebeom for your wonderful contribution! I put some comments — could you take them into account please?

kimdaebeom commented:
Thanks @LimHyungTae for your feedback! I have addressed all the comments and updated the code accordingly. Let me know if there’s anything else to refine. :+1:

LimHyungTae reviewed (commented):
Thanks for your effort! Could you finally deal with these things?

kimdaebeom commented:
@LimHyungTae   I have reflected your feedback. Can you check the commit again please?

LimHyungTae commented:
Oh man, why did you also change the name of the executable?

kimdaebeom commented:
I apologize for my mistake. I should have only renamed the package, considering that this ROS package will include multiple applications including SLAM. I would appreciate it if you could review it once again. Please let me know if there are any additional modifications needed.
